internal/controller/serverbios_controller.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		return err
+	}
+	if server.Spec.BIOSSettingsRef == (corev1.LocalObjectReference{}) {
+		return nil
+	}
+	if server.Spec.BIOSSettingsRef.Name != serverBIOS.Name {
+		return nil
+	}
+	return r.patchBIOSSettingsRef(ctx, log, &server, corev1.LocalObjectReference{})
+}
+
+// Reconciliation flow for ServerBIOS:
+//  1. Ensure finalizer is set on the object
+//  2. Ensure info about current BIOS version and settings is not outdated, otherwise:
+//     2.2. Invoke scan job
+//     2.2. Exit if no discrepancy is found
+//  3. Ensure referred server is in Available state
+//  4. Ensure desired and current BIOS versions match, otherwise:
+//     4.1. Exit if some task is already in progress
+//     4.2. Invoke BIOS version update job
+//  5. Ensure desired and current BIOS settings match, otherwise:
+//     5.1. Exit if some task is already in progress
+//     5.2. Invoke BIOS settings update job
+func (r *ServerBIOSReconciler) reconcile(
+	ctx context.Context,
+	log logr.Logger,
+	serverBIOS *metalv1alpha1.ServerBIOS,
+) (ctrl.Result, error) {
+	if modified, err := clientutils.PatchEnsureFinalizer(ctx, r.Client, serverBIOS, serverBIOSFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("ensured finalizer has been added")
+
+	updateRequired, err := r.reconcileScan(ctx, log, serverBIOS)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if updateRequired {
+		return r.reconcileUpdate(ctx, log, serverBIOS)
+
+	}
+	return ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
+}
+
+func (r *ServerBIOSReconciler) reconcileScan(
+	ctx context.Context,
+	log logr.Logger,
+	serverBIOS *metalv1alpha1.ServerBIOS,
+) (bool, error) {
+	log.V(1).Info("invoking scan job")
+	result, err := r.TaskRunnerClient.Scan(ctx, serverBIOS.Name)
+	if err != nil {
+		return false, err
+	}
+	serverBIOSBase := serverBIOS.DeepCopy()
+	versionUpdateRequired := serverBIOS.Spec.BIOS.Version != "" && (serverBIOS.Spec.BIOS.Version != result.Version)
+	if !versionUpdateRequired {
+		serverBIOS.Status.BIOS.Version = result.Version
+	}
+	settingsUpdateRequired := len(serverBIOS.Spec.BIOS.Settings) != 0 &&
+		!cmp.Equal(serverBIOS.Spec.BIOS.Settings, result.Settings)
+	if !settingsUpdateRequired {
+		serverBIOS.Status.BIOS.Settings = result.Settings
+	}
+	updateRequired := versionUpdateRequired || settingsUpdateRequired
+	return updateRequired, r.Status().Patch(ctx, serverBIOS, client.MergeFrom(serverBIOSBase))
+}
+
+func (r *ServerBIOSReconciler) reconcileUpdate(
+	ctx context.Context,
+	log logr.Logger,
+	serverBIOS *metalv1alpha1.ServerBIOS,
+) (ctrl.Result, error) {
+	server, err := r.getReferredServer(ctx, log, serverBIOS.Spec.ServerRef.Name)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// if referred server is not in Available state - stop reconciliation
+	if server.Status.State != metalv1alpha1.ServerStateAvailable {
+		return ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
+	}
```

Comment:
do we need to stop the server reconciliation in the meantime?

Reply:
Yes. Bc both BIOS version and settings update will lead to server reboot. So for now we decided that we'll work only with servers which are in "Available" state.

Reply:
> ok but during a serverBios update, someone could still claim the server. Any available server can be claimed

@stefanhipfel that's a good point. However, I do not really like the idea that the serverBIOS controller would change server's state to exclude it from reconciliation. For number of reasons the server's "owner" might want to postpone the update. Especially in terms of version upgrade.
I could add the "Maintenance" state for server and update the controller, so it will check whether server's state is "Maintenance" instead of "Available".

Reply:
@aobort yes i think so as well, that the serverBIOS should not change a server's state. 

Alternative idea: the server reconciler checks on the serverBIOS state.